server/data/data-jmap/src/main/java/org/apache/james/jmap/api/change/MailboxChange.java

```diff
@@ -92,60 +100,201 @@ public int getValue() {
-        return new MailboxChange(accountId, state, date, created, updated, destroyed);
+    @FunctionalInterface
+    interface RequiredAccountId {
+        RequiredState accountId(AccountId accountId);
     }
 
-    public static Optional<MailboxChange> fromEvent(Event event) {
+    @FunctionalInterface
+    interface RequiredState {
+        RequiredDate state(State state);
+    }
+
+    @FunctionalInterface
+    interface RequiredDate {
+        Builder date(ZonedDateTime date);
+    }
+
+    public static class Builder {
+        private final AccountId accountId;
+        private final State state;
+        private final ZonedDateTime date;
+        private boolean delegated;
+        private Optional<List<MailboxId>> created;
+        private Optional<List<MailboxId>> updated;
+        private Optional<List<MailboxId>> destroyed;
+
+        private Builder(AccountId accountId, State state, ZonedDateTime date) {
+            this.accountId = accountId;
+            this.state = state;
+            this.date = date;
+            this.created = Optional.empty();
+            this.updated = Optional.empty();
+            this.destroyed = Optional.empty();
+        }
+
+        public Builder delegated() {
+            this.delegated = true;
+            return this;
+        }
+
+        public Builder created(List<MailboxId> created) {
+            this.created = Optional.of(created);
+            return this;
+        }
+
+        public Builder updated(List<MailboxId> updated) {
+            this.updated = Optional.of(updated);
+            return this;
+        }
+
+        public Builder destroyed(List<MailboxId> destroyed) {
+            this.destroyed = Optional.of(destroyed);
+            return this;
+        }
+
+        public MailboxChange build() {
+            return new MailboxChange(accountId, state, date, delegated, created.orElse(ImmutableList.of()), updated.orElse(ImmutableList.of()), destroyed.orElse(ImmutableList.of()));
+        }
+    }
+
+    public static RequiredAccountId builder() {
+        return accountId -> state -> date -> new Builder(accountId, state, date);
+    }
+
+    public static Builder created(AccountId accountId, State state, ZonedDateTime date, List<MailboxId> created) {
+        return MailboxChange.builder()
+            .accountId(accountId)
+            .state(state)
+            .date(date)
+            .created(created);
+    }
+
+    public static Builder updated(AccountId accountId, State state, ZonedDateTime date, List<MailboxId> updated) {
+        return MailboxChange.builder()
+            .accountId(accountId)
+            .state(state)
+            .date(date)
+            .updated(updated);
+    }
+
+    public static Builder destroyed(AccountId accountId, State state, ZonedDateTime date, List<MailboxId> destroyed) {
+        return MailboxChange.builder()
+            .accountId(accountId)
+            .state(state)
+            .date(date)
+            .destroyed(destroyed);
+    }
```

Comment:
I find it weird to "hide" builder calls behind their multi-argument cryptic method calls.

Do we really need these created updated destroyed factory methods?